[JSC][armv7] Support for lowering B3 atomics to Air for armv7

eugeneia opened the pull request:
#### fb0137cc6ec2d561565607868e0986c7c6079661
<pre>
[JSC][armv7] Support for lowering B3 atomics to Air for armv7
<a href="https://bugs.webkit.org/show_bug.cgi?id=282433">https://bugs.webkit.org/show_bug.cgi?id=282433</a>

Reviewed by Justin Michaud.

Implements various atomic operations for OMG on armv7.

* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::loadAcq32):
(JSC::MacroAssemblerARMv7::loadAcq8):
(JSC::MacroAssemblerARMv7::loadAcq8SignedExtendTo32):
(JSC::MacroAssemblerARMv7::loadAcq16):
(JSC::MacroAssemblerARMv7::loadAcq16SignedExtendTo32):
(JSC::MacroAssemblerARMv7::storeRel32):
(JSC::MacroAssemblerARMv7::storeRel8):
(JSC::MacroAssemblerARMv7::storeRel16):
* Source/JavaScriptCore/b3/B3LowerInt64.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/testb3_failingArmV7Tests.inc:

Canonical link: <a href="https://commits.webkit.org/286082@main">https://commits.webkit.org/286082@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73b1c7c501f3b6cf7b02c908c0c00271ef4f400f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74788 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54218 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27605 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79211 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26026 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63354 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2003 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58743 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17025 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77855 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48895 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64268 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39137 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46158 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21765 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24359 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67925 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22110 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80701 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74046 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2106 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1258 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67003 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2254 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64286 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66296 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16462 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10237 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8392 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96317 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2071 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21058 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2099 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3019 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2106 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
```diff
@@ -2904,16 +2960,78 @@ class LowerToAir {
+        // We need a CAS loop or a LL/SC loop. Using prepare/attempt jargon, we want:
+        //
+        // Block #reloop:
+        //     Prepare
```
I guess this is a general issue, but a tight loop like this will break the JSC watchdog, which might make fuzzing harder. There are other places that are affected too, but we should keep this section in mind if we ever get around to fixing that.

justinmichaud reviewed (approved):
r=me

aoikonomopoulos commented on Source/JavaScriptCore/b3/B3LowerInt64.cpp:
```diff
@@ -1184,6 +1184,41 @@ class LowerInt64 {
+            auto address = atomic->child(2);
+            auto stitchedExpected = valueLoHi(expectedValue.first, expectedValue.second, m_index);
+            auto stitchedNew = valueLoHi(newValue.first, newValue.second, m_index);
+            auto stitched = insert<AtomicValue>(m_index, atomic->opcode(), m_origin, atomic->accessWidth(), stitchedExpected, stitchedNew, address);
```
Can we copy the other fields of `AtomicValue` (specifically those of its superclass `MemoryValue`) from `m_value`? Similar to what we do for `Load` in this file.

eugeneia committed:
[JSC][armv7] Support for lowering B3 atomics to Air for armv7
https://bugs.webkit.org/show_bug.cgi?id=282433

Reviewed by Justin Michaud.

Implements various atomic operations for OMG on armv7.

* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::loadAcq32):
(JSC::MacroAssemblerARMv7::loadAcq8):
(JSC::MacroAssemblerARMv7::loadAcq8SignedExtendTo32):
(JSC::MacroAssemblerARMv7::loadAcq16):
(JSC::MacroAssemblerARMv7::loadAcq16SignedExtendTo32):
(JSC::MacroAssemblerARMv7::storeRel32):
(JSC::MacroAssemblerARMv7::storeRel8):
(JSC::MacroAssemblerARMv7::storeRel16):
* Source/JavaScriptCore/b3/B3LowerInt64.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/testb3_failingArmV7Tests.inc:

Canonical link: https://commits.webkit.org/286082@main
